### DISCUSSION TITLE
코드의 평가 과정에서, 매개변수, 지역변수, arguments 객체에 들어가는 값들을 설명해주세요.

### DISCUSSION THREAD

**Discussion by pipisebastian** [publdaze/js-study#76] [실행컨텍스트]
```js
function add(a) {
		var b = 2;
		let c = 3;
    console.log(a+b+c);
}

add(1);
```
> 코드의 평가 과정에서, 매개변수, 지역변수, arguments 객체에 들어가는 값들을 설명해주세요. 

> 코드의 실행 과정에서, 매개변수, 지역변수, arguments 객체에 들어가는 값들을 설명해주세요.

> console.log가 실행되는 과정을, 실행컨텍스트와 관련지어서 설명해주세요

**Comment by pipisebastian** [comment]
> 코드의 평가 과정에서, 매개변수, 지역변수, arguments 객체에 들어가는 값들을 설명해주세요.

- a : undefined
- b : undefined
- c : 값 x, uninitilalized
- arguments : {0:1, length:1, callee:add}


> 코드의 실행 과정에서, 매개변수, 지역변수, arguments 객체에 들어가는 값들을 설명해주세요.

- a : 1
- b : 2
- c : 3
- arguments : {0:1, length:1, callee:add}

> console.log가 실행되는 과정을, 실행컨텍스트와 관련지어서 설명해주세요.

- console 식별자를 스코프체인에서 검색 → add 컨텍스트에 없음 → 상위 스코프(전역) 검색
- 전역 컨텍스트 → enviornment record → 객체 환경 레코드 → bindingObject를 통해, 전역 객체 window에서 찾음
- console 객체에서 log 메서드 찾음!
- a+b+c에서 각각의 식별자 찾기! 싹다 add 컨텍스트안에 있음! 끝!
- 표현식 a+b+c가 평가된 값을 console.log메서드로 전달!

**Comment by publdaze** [comment]
> ```js
> function add(a) {
>     var b = 2;
>     let c = 3;
>     console.log(a+b+c);
> }
> 
> add(1);
> ```
> - 코드의 평가 과정에서, 매개변수, 지역변수, arguments 객체에 들어가는 값들을 설명해주세요.
- add가 호출되면 add 함수의 실행 컨텍스트가 생성되고, 환경 레코드에 매개 변수, 지역 변수, arguments 객체가 등록되고 관리 됨
- 매개변수 a의 경우 평가 과정에서 지역 변수와 동일하게 취급되어 암묵적으로 매개변수가 선언되고 undefined로 초기화 됨.
- 지역변수 b의 경우 var로 선언되어 있으므로 호이스팅이 일어나며, 선언과 초기화 단계가 동시에 일어남. 따라서 b는 undefined가 됨
- 지역변수 c의 경우 const로 선언되어 있음 const도 호이스팅이 일어나지만 선언단계가 코드 평가될 때 먼저 일어나고 초기화단계는 실행단계에서 선언문이 실행될 때 초기화 단계를 거침. 따라서 선언 단계와 초기화 단계 사이에는 일시적 사각지대가 존재하고 초기화되지 않은 상태로 남아있음. 따라서 c 변수에는 접근이 불가 함
- arguments 객체에는 함수 호출문의 인수가 프로퍼티로 보관되고, 유사 배열처럼 인수가 들어오는 순서대로 인덱스를 key로 가지고 인수 값은 value가 됨. 그 외에도 길이를 나타내는 length 프로퍼티와 함수 자신을 가리티는 callee프로퍼티 존재함. 따라서 평가 과정의 arguments 값은 { 0: 1, length: 1, callee: add} 가 됨

> - 코드의 실행 과정에서, 매개변수, 지역변수, arguments 객체에 들어가는 값들을 설명해주세요.
- 코드가 순차적으로 실행되면서 매개변수에 인수가 할당되고, b, c 변수 할당문이 실행되어 각각 2, 3이 할당되고, arguments 객체는 변경없이 평가단계와 동일한 값이 유지됨.
> - console.log가 실행되는 과정을, 실행컨텍스트와 관련지어서 설명해주세요
- add 함수 코드가 실행되면서 console.log 함수 호출문을 만나면 먼저 console 식별자를 검색하는 과정을 거침
- console은 전역 객체의 프로퍼티로 스코프체인의 종점에 존재하므로 현재 실행 중인 실행 컨텍스트의 렉시컬 환경부터 시작해 스코프 체인을 타고 외부 렉시컬 환경으로 이동하다 전역 렉시컬 환경에 도달했을 때 식별자가 결정됨
-  console 객체에서 log 메서드 검색
- 인수로 전달되는 a, b, c는 console.log가 호출되어 실행되면 매개변수에 할당되고 연산이 수행 된 후 출력됨
